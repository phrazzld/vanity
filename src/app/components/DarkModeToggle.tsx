'use client';

import React, { useState, useEffect, useRef } from 'react';
import { useTheme } from '../context/ThemeContext';

export type DarkModeToggleProps = {
  /**
   * The size of the toggle button
   * @default "medium"
   */
  size?: 'small' | 'medium' | 'large';
  /**
   * Additional CSS classes to apply to the button
   */
  className?: string;
  /**
   * Optional click handler
   * In most cases, you should let the component use the ThemeContext's toggleDarkMode
   */
  onClick?: () => void;
  /**
   * Force a specific mode (overrides ThemeContext)
   */
  forcedMode?: 'light' | 'dark';
};

const DarkModeToggle = React.memo(function DarkModeToggle({
  size = 'medium',
  className = '',
  onClick,
  forcedMode,
}: DarkModeToggleProps) {
  const { isDarkMode: contextDarkMode, toggleDarkMode } = useTheme();
  const buttonRef = useRef<HTMLButtonElement>(null);
  const [hasBeenVisible, setHasBeenVisible] = useState(false);

  /**
   * Animation State Management
   *
   * The component uses a multi-layered animation system:
   * 1. isAnimating: Tracks whether the toggle animation is currently running
   *    - Prevents spam clicking during the 200ms animation duration
   *    - Triggers the subtle scale pulse animation when true
   *
   * 2. hasBeenVisible: Lazy-loads animations using Intersection Observer
   *    - Defers animation classes until component is visible in viewport
   *    - Improves initial page load performance by avoiding unnecessary GPU acceleration
   *    - Once visible, animations are permanently enabled
   *
   * 3. GPU Acceleration: Uses transform-gpu class after visibility
   *    - Forces GPU acceleration for smoother 60fps animations
   *    - Only applied after component has been seen to reduce memory usage
   *
   * Animation Flow:
   * - User clicks → isAnimating = true → icon scales down and back with subtle pulse
   * - After 200ms → isAnimating = false → ready for next interaction
   * - Theme transition CSS handles color/background changes separately (300ms)
   */
  const [isAnimating, setIsAnimating] = useState(false);
  const handleToggle = () => {
    if (isAnimating) return; // Prevent spam clicking
    setIsAnimating(true);
    (onClick || toggleDarkMode)();
    setTimeout(() => setIsAnimating(false), 200);
  };

  // Intersection observer to defer animations until visible
  useEffect(() => {
    if (!buttonRef.current || hasBeenVisible) return;

    const observer = new IntersectionObserver(
      entries => {
        entries.forEach(entry => {
          if (entry.isIntersecting) {
            setHasBeenVisible(true);
            observer.disconnect();
          }
        });
      },
      { threshold: 0.1 }
    );

    observer.observe(buttonRef.current);

    return () => observer.disconnect();
  }, [hasBeenVisible]);

  // Preload both icon states to prevent flicker
  useEffect(() => {
    // Create hidden elements to force browser to parse both SVGs
    const preloadContainer = document.createElement('div');
    preloadContainer.style.position = 'absolute';
    preloadContainer.style.width = '0';
    preloadContainer.style.height = '0';
    preloadContainer.style.overflow = 'hidden';
    preloadContainer.setAttribute('aria-hidden', 'true');

    // Sun icon
    const sunSvg = document.createElementNS('http://www.w3.org/2000/svg', 'svg');
    sunSvg.innerHTML =
      '<path d="M12 3v1m0 16v1m9-9h-1M4 12H3m15.364 6.364l-.707-.707M6.343 6.343l-.707-.707m12.728 0l-.707.707M6.343 17.657l-.707.707M16 12a4 4 0 11-8 0 4 4 0 018 0z"/>';

    // Moon icon
    const moonSvg = document.createElementNS('http://www.w3.org/2000/svg', 'svg');
    moonSvg.innerHTML =
      '<path d="M20.354 15.354A9 9 0 018.646 3.646 9.003 9.003 0 0012 21a9.003 9.003 0 008.354-5.646z"/>';

    preloadContainer.appendChild(sunSvg);
    preloadContainer.appendChild(moonSvg);
    document.body.appendChild(preloadContainer);

    return () => {
      if (preloadContainer.parentNode) {
        document.body.removeChild(preloadContainer);
      }
    };
  }, []);

  // Use forcedMode if provided, otherwise use the context value
  const isDarkMode = forcedMode === undefined ? contextDarkMode : forcedMode === 'dark';

  // Determine icon size based on the size prop
  const iconSize = {
    small: 'h-4 w-4',
    medium: 'h-5 w-5',
    large: 'h-6 w-6',
  }[size];

  // Determine button padding based on size
  const buttonPadding = {
    small: 'p-1.5',
    medium: 'p-2',
    large: 'p-2.5',
  }[size];

  // Base classes for the button
  const baseClasses =
    'dark-mode-toggle relative flex items-center justify-center bg-transparent rounded-full ' +
    'transition-all duration-300 ease-elegant-entrance ' +
    'hover:bg-gray-100 dark:hover:bg-gray-800 hover:shadow-md hover:scale-105 ' +
    'active:scale-[0.98] focus:outline-none focus:ring-2 focus:ring-primary-500 focus:ring-offset-2';

  return (
    <button
      ref={buttonRef}
      className={`${baseClasses} ${buttonPadding} ${className}`}
      onClick={handleToggle}
      aria-label={isDarkMode ? 'Switch to light mode' : 'Switch to dark mode'}
    >
      <div className="relative w-full h-full flex items-center justify-center">
        {isDarkMode ? (
          <svg
            xmlns="http://www.w3.org/2000/svg"
            className={`${iconSize} ${hasBeenVisible && isAnimating ? 'animate-icon-switch' : ''} text-yellow-500 ${hasBeenVisible ? 'transform-gpu' : ''}`}
            fill="none"
            viewBox="0 0 24 24"
            stroke="currentColor"
          >
            <defs>
              <linearGradient id="sun-gradient" x1="0%" y1="0%" x2="100%" y2="100%">
                <stop offset="0%" stopColor="#fbbf24" />
                <stop offset="100%" stopColor="#f59e0b" />
              </linearGradient>
            </defs>
            <path
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth={2}
              fill="url(#sun-gradient)"
              d="M12 3v1m0 16v1m9-9h-1M4 12H3m15.364 6.364l-.707-.707M6.343 6.343l-.707-.707m12.728 0l-.707.707M6.343 17.657l-.707.707M16 12a4 4 0 11-8 0 4 4 0 018 0z"
            />
          </svg>
        ) : (
          <svg
            xmlns="http://www.w3.org/2000/svg"
            className={`${iconSize} ${hasBeenVisible && isAnimating ? 'animate-icon-switch' : ''} text-slate-700 dark:text-slate-300 ${hasBeenVisible ? 'transform-gpu' : ''}`}
            fill="none"
            viewBox="0 0 24 24"
            stroke="currentColor"
          >
            <path
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth={2}
              d="M20.354 15.354A9 9 0 018.646 3.646 9.003 9.003 0 0012 21a9.003 9.003 0 008.354-5.646z"
            />
          </svg>
        )}
      </div>
    </button>
  );
});

export default DarkModeToggle;
