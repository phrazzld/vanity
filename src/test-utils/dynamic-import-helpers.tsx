/**
 * Test helpers for Next.js dynamic imports
 *
 * These utilities help test components that use Next.js dynamic()
 * for code splitting and lazy loading.
 */

import React from 'react';
import { waitFor } from '@testing-library/react';
import preloadAll from 'jest-next-dynamic';

/**
 * Preload all dynamic imports before running tests
 * Call this in beforeAll() hooks for test suites that use dynamic imports
 */
export async function preloadDynamicImports(): Promise<void> {
  await preloadAll();
}

/**
 * Wait for a dynamic component to load and render
 * Useful when testing components that show loading states
 */
export async function waitForDynamicComponent(
  getElement: () => HTMLElement | null,
  options?: { timeout?: number }
): Promise<void> {
  await waitFor(
    () => {
      const element = getElement();
      expect(element).toBeInTheDocument();
    },
    { timeout: options?.timeout || 5000 }
  );
}

/**
 * Mock for next/dynamic that immediately renders the component
 * without async loading behavior - useful for unit tests
 */
export function createSyncDynamicMock() {
  return jest.fn().mockImplementation((importFunc: () => Promise<any>, _options?: any) => {
    // If there's a loading component, we'll ignore it for sync tests
    // Return a component that renders immediately
    const Component = jest.fn().mockImplementation((props: any) => {
      const MockComponent = require(
        importFunc.toString().match(/import\(['"](.+)['"]\)/)?.[1] || ''
      );
      return MockComponent.default ? MockComponent.default(props) : MockComponent(props);
    });

    return Component;
  });
}

/**
 * Mock for next/dynamic that simulates async loading
 * Provides better control over loading states in tests
 */
export function createAsyncDynamicMock() {
  return jest.fn().mockImplementation((importFunc: () => Promise<any>, options?: any) => {
    const LoadingComponent = options?.loading || (() => null);

    return function MockDynamicComponent(props: any) {
      const [isLoading, setIsLoading] = React.useState(true);
      const [Component, setComponent] = React.useState<any>(null);

      React.useEffect(() => {
        // Simulate async import
        void importFunc().then(module => {
          setComponent(() => module.default || module);
          setIsLoading(false);
        });
      }, []);

      if (isLoading) {
        return <LoadingComponent />;
      }

      return Component ? <Component {...props} /> : null;
    };
  });
}
