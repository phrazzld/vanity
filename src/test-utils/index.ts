// src/test-utils/index.ts
import React, { useState } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ThemeContext } from '@/app/context/ThemeContext';
import userEvent from '@testing-library/user-event';

// ===========================================================================
// Theme Provider Test Setup
// ===========================================================================

/**
 * A custom theme provider for tests that doesn't use browser APIs
 */
export function TestThemeProvider({
  children,
  initialTheme = 'light',
}: {
  children: React.ReactNode;
  initialTheme?: 'light' | 'dark';
}) {
  const [isDarkMode, setIsDarkMode] = useState(initialTheme === 'dark');

  const toggleDarkMode = () => setIsDarkMode(prev => !prev);

  return React.createElement(
    ThemeContext.Provider,
    { value: { isDarkMode, toggleDarkMode } },
    React.createElement(
      'div',
      { 
        'data-testid': 'theme-provider',
        'data-theme': isDarkMode ? 'dark' : 'light'
      },
      children
    )
  );
}

/**
 * Custom render method that includes the TestThemeProvider wrapper
 * Simplified from https://testing-library.com/docs/react-testing-library/setup
 */
type CustomRenderOptions = Omit<RenderOptions, 'wrapper'> & {
  themeMode?: 'light' | 'dark';
  // Add other custom render options here as needed
};

export function renderWithTheme(
  ui: React.ReactElement,
  options?: CustomRenderOptions
) {
  const { themeMode = 'light', ...renderOptions } = options || {};
  
  return render(ui, {
    wrapper: ({ children }) => (
      React.createElement(TestThemeProvider, { initialTheme: themeMode }, children)
    ),
    ...renderOptions,
  });
}

// Setup user-event with standard configuration
export function setupUser(options = { delay: null }) {
  return userEvent.setup(options);
}

// ===========================================================================
// Mock Response Helpers
// ===========================================================================

// Mock Fetch Response Helper
export function createMockResponse<T>(data: T, status = 200, statusText = 'OK') {
  return {
    ok: status >= 200 && status < 300,
    status,
    statusText,
    json: jest.fn().mockResolvedValue(data),
    text: jest.fn().mockResolvedValue(JSON.stringify(data)),
  };
}

// Create a failed response
export function createMockErrorResponse(
  status = 500,
  statusText = 'Server Error',
  message = 'An error occurred'
) {
  return {
    ok: false,
    status,
    statusText,
    json: jest.fn().mockResolvedValue({ error: message }),
    text: jest.fn().mockResolvedValue(JSON.stringify({ error: message })),
  };
}

// Mocks the fetch for testing API calls
export function mockFetch<T>(data: T, status = 200) {
  const mockResponse = createMockResponse(data, status);
  const fetchMock = jest.fn().mockResolvedValue(mockResponse);
  // Cast window to unknown first to avoid TypeScript errors
  (window as unknown as { fetch: typeof fetchMock }).fetch = fetchMock;
  return { mockResponse, mockFetch: fetchMock };
}

// ===========================================================================
// Test Utilities
// ===========================================================================

// Helper to wait for a condition
export function waitForCondition(
  callback: () => boolean | Promise<boolean>,
  { timeout = 1000, interval = 50 } = {}
): Promise<void> {
  return new Promise((resolve, reject) => {
    const startTime = Date.now();

    const checkCondition = async () => {
      const result = await callback();
      if (result) {
        resolve();
        return;
      }

      const elapsedTime = Date.now() - startTime;
      if (elapsedTime > timeout) {
        reject(new Error(`Timed out waiting for condition after ${timeout}ms`));
        return;
      }

      setTimeout(checkCondition, interval);
    };

    checkCondition();
  });
}

// Hook testing with theme context
import { renderHook, RenderHookOptions, RenderHookResult } from '@testing-library/react';

/**
 * Custom renderHook method that includes the ThemeProvider wrapper
 */
export function renderHookWithTheme<Result, Props>(
  hook: (props: Props) => Result,
  options?: Omit<RenderHookOptions<Props>, 'wrapper'> & {
    themeMode?: 'light' | 'dark';
  }
): RenderHookResult<Result, Props> {
  const { themeMode = 'light', ...renderOptions } = options || {};
  
  return renderHook(hook, {
    wrapper: ({ children }) => (
      React.createElement(TestThemeProvider, { initialTheme: themeMode }, children)
    ),
    ...renderOptions,
  });
}

// Export everything from testing library for convenience
export * from '@testing-library/react';
export { userEvent };
