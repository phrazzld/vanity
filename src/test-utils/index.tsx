// src/test-utils/index.tsx
import React from 'react';
import { render } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

// ===========================================================================
// Theme Provider Test Setup
// ===========================================================================

/**
 * A simple test wrapper that sets a data-theme attribute for testing
 * Note: Theme is now managed by Zustand store, so we don't need a provider
 */
export function TestThemeProvider({
  children,
  initialTheme = 'light',
}: {
  children: React.ReactNode;
  initialTheme?: 'light' | 'dark';
}) {
  return <div data-theme={initialTheme}>{children}</div>;
}

/**
 * Custom render method that includes the TestThemeProvider wrapper
 * Simplified from https://testing-library.com/docs/react-testing-library/setup
 */
type CustomRenderOptions = Omit<RenderOptions, 'wrapper'> & {
  themeMode?: 'light' | 'dark';
  // Add other custom render options here as needed
};

export function renderWithTheme(ui: React.ReactElement, options?: CustomRenderOptions) {
  const { themeMode = 'light', ...renderOptions } = options || {};

  return render(ui, {
    wrapper: ({ children }) => (
      <TestThemeProvider initialTheme={themeMode}>{children}</TestThemeProvider>
    ),
    ...renderOptions,
  });
}

// Setup user-event with standard configuration
export function setupUser(options = { delay: null }) {
  return userEvent.setup(options);
}

// ===========================================================================
// Mock Response Helpers
// ===========================================================================

/**
 * Type for the mock response returned by createMockResponse
 */
export type MockResponse<T> = {
  ok: boolean;
  status: number;
  statusText: string;
  json: jest.Mock<Promise<T>>;
  text: jest.Mock<Promise<string>>;
};

/**
 * Type for the return value of mockFetch
 */
export type MockFetchReturn<T> = {
  mockResponse: MockResponse<T>;
  mockFetch: jest.Mock<Promise<MockResponse<T>>>;
};

// Mock Fetch Response Helper
export function createMockResponse<T>(data: T, status = 200, statusText = 'OK'): MockResponse<T> {
  return {
    ok: status >= 200 && status < 300,
    status,
    statusText,
    json: jest.fn().mockResolvedValue(data) as jest.Mock<Promise<T>>,
    text: jest.fn().mockResolvedValue(JSON.stringify(data)) as jest.Mock<Promise<string>>,
  };
}

// Create a failed response
export function createMockErrorResponse(
  status = 500,
  statusText = 'Server Error',
  message = 'An error occurred'
): MockResponse<{ error: string }> {
  return {
    ok: false,
    status,
    statusText,
    json: jest.fn().mockResolvedValue({ error: message }) as jest.Mock<Promise<{ error: string }>>,
    text: jest.fn().mockResolvedValue(JSON.stringify({ error: message })) as jest.Mock<
      Promise<string>
    >,
  };
}

// Mocks the fetch for testing API calls
export function mockFetch<T>(data: T, status = 200): MockFetchReturn<T> {
  const mockResponse = createMockResponse(data, status);
  const fetchMock = jest.fn().mockResolvedValue(mockResponse) as jest.Mock<
    Promise<MockResponse<T>>
  >;
  // Cast window to unknown first to avoid TypeScript errors
  (window as unknown as { fetch: typeof fetchMock }).fetch = fetchMock;
  return { mockResponse, mockFetch: fetchMock };
}

// ===========================================================================
// Test Utilities
// ===========================================================================

// Helper to wait for a condition
export function waitForCondition(
  callback: () => boolean | Promise<boolean>,
  { timeout = 1000, interval = 50 } = {}
): Promise<void> {
  return new Promise((resolve, reject) => {
    const startTime = Date.now();

    const checkCondition = async () => {
      const result = await callback();
      if (result) {
        resolve();
        return;
      }

      const elapsedTime = Date.now() - startTime;
      if (elapsedTime > timeout) {
        reject(new Error(`Timed out waiting for condition after ${timeout}ms`));
        return;
      }

      setTimeout(checkCondition, interval);
    };

    checkCondition();
  });
}

// Hook testing with theme context
import { renderHook } from '@testing-library/react';
import type { RenderHookOptions, RenderHookResult } from '@testing-library/react';

/**
 * Custom renderHook method that includes the ThemeProvider wrapper
 */
export function renderHookWithTheme<Result, Props>(
  hook: (_props: Props) => Result,
  options?: Omit<RenderHookOptions<Props>, 'wrapper'> & {
    themeMode?: 'light' | 'dark';
  }
): RenderHookResult<Result, Props> {
  // _initialProps would be used by the renderHook function internally
  const { themeMode = 'light', ...renderOptions } = options || {};

  return renderHook(hook, {
    wrapper: ({ children }) => (
      <TestThemeProvider initialTheme={themeMode}>{children}</TestThemeProvider>
    ),
    ...renderOptions,
  });
}

// ===========================================================================
// Snapshot Testing Utilities
// ===========================================================================

/**
 * Creates a snapshot test with consistent formatting
 *
 * @param ui The React element to render
 * @param options The render options, including theme mode
 * @returns The rendered component for snapshot testing
 */
export function createComponentSnapshot(ui: React.ReactElement, options?: CustomRenderOptions) {
  const { container } = renderWithTheme(ui, options);

  // Remove whitespace for cleaner snapshots
  const cleanHtml = container.innerHTML
    .replace(/\s{2,}/g, ' ')
    .replace(/\n/g, '')
    .trim();

  return {
    container,
    cleanHtml,
  };
}

/**
 * Creates a responsive snapshot for multiple viewport sizes
 * Sets the viewport size using Jest's mocked window object
 */
export function createResponsiveSnapshots(
  ui: React.ReactElement,
  viewportSizes = ['mobile', 'tablet', 'desktop'],
  options?: CustomRenderOptions
) {
  const originalInnerWidth = window.innerWidth;
  const originalInnerHeight = window.innerHeight;

  const viewportMap: Record<string, { width: number; height: number }> = {
    mobile: { width: 390, height: 844 },
    tablet: { width: 768, height: 1024 },
    desktop: { width: 1280, height: 800 },
  };

  const snapshots: Record<string, string> = {};

  // Generate snapshot for each viewport size
  viewportSizes.forEach(size => {
    // Set viewport size
    const viewport = viewportMap[size];
    if (!viewport) return;

    // Use temporary variables to avoid TypeScript's unsafe assignment error
    const viewportWidth = viewport.width;
    const viewportHeight = viewport.height;

    // Use explicit type for configurable object to avoid unsafe assignment
    const widthConfig: PropertyDescriptor = {
      writable: true,
      configurable: true,
      value: viewportWidth,
    };
    const heightConfig: PropertyDescriptor = {
      writable: true,
      configurable: true,
      value: viewportHeight,
    };

    Object.defineProperty(window, 'innerWidth', widthConfig);
    Object.defineProperty(window, 'innerHeight', heightConfig);

    // Create window.matchMedia mock for the current viewport
    // Use width value directly to avoid unsafe assignment
    const width = viewport.width;
    const isDarkMode = options?.themeMode === 'dark';

    // Create type-safe query matcher function
    const matchesQuery = (query: string): boolean => {
      return (
        query.includes(`(min-width: ${width}px)`) ||
        (query.includes('(prefers-color-scheme: dark)') && isDarkMode)
      );
    };

    const mockMatchMedia = jest.fn().mockImplementation((query: string) => {
      return {
        matches: matchesQuery(query),
        media: query,
        onchange: null,
        addListener: jest.fn(),
        removeListener: jest.fn(),
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      };
    });

    // Type assertion to avoid TypeScript errors
    window.matchMedia = mockMatchMedia as typeof window.matchMedia;

    // Render the component
    const { cleanHtml } = createComponentSnapshot(ui, options);
    snapshots[size] = cleanHtml;
  });

  // Restore original viewport size
  // Use temporary variables to avoid unsafe assignment
  const originalWidth = originalInnerWidth;
  const originalHeight = originalInnerHeight;

  // Use explicit type for configurable object to avoid unsafe assignment
  const originalWidthConfig: PropertyDescriptor = {
    writable: true,
    configurable: true,
    value: originalWidth,
  };
  const originalHeightConfig: PropertyDescriptor = {
    writable: true,
    configurable: true,
    value: originalHeight,
  };

  Object.defineProperty(window, 'innerWidth', originalWidthConfig);
  Object.defineProperty(window, 'innerHeight', originalHeightConfig);

  return snapshots;
}

// ===========================================================================
// Accessibility Testing Utilities
// ===========================================================================

// Import and export accessibility testing helpers
export * from './a11y-helpers';

// Custom matcher for testing theme differences
declare global {
  namespace jest {
    interface Matchers<R> {
      toMatchThemeSnapshots: (_comparison: unknown) => R;
    }
  }
}

// Add the matcher to Jest's expect
expect.extend({
  toMatchThemeSnapshots(received, comparison) {
    const pass = received.cleanHtml !== comparison.cleanHtml;

    if (pass) {
      return {
        message: () => `Expected theme snapshots to be different and they are.`,
        pass: true,
      };
    } else {
      return {
        message: () => `Expected theme snapshots to be different, but they are the same.`,
        pass: false,
      };
    }
  },
});

// Export everything from testing library for convenience
export * from '@testing-library/react';
export { userEvent };
