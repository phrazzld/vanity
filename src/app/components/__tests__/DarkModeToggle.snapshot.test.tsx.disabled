/**
 * DarkModeToggle Component Snapshot Tests
 *
 * This file demonstrates snapshot testing patterns for the DarkModeToggle component.
 * It tests various component states, sizes, and themes.
 */

import { createComponentSnapshot } from '@/test-utils';
import DarkModeToggle from '../DarkModeToggle';

describe('DarkModeToggle Component Snapshots', () => {
  describe('Size Variants', () => {
    it('renders in different sizes', () => {
      const sizes = ['small', 'medium', 'large'] as const;

      sizes.forEach(size => {
        const { cleanHtml } = createComponentSnapshot(<DarkModeToggle size={size} />);

        expect(cleanHtml).toMatchSnapshot(`size: ${size}`);
      });
    });
  });

  describe('Theme Variants', () => {
    it('renders light and dark mode icons correctly', () => {
      // Test with forced light mode
      const lightRender = createComponentSnapshot(<DarkModeToggle forcedMode="light" />);

      // Test with forced dark mode
      const darkRender = createComponentSnapshot(<DarkModeToggle forcedMode="dark" />);

      // Verify toggle shows different icons for light/dark
      expect(lightRender).toMatchThemeSnapshots(darkRender);

      // Also snapshot each theme
      expect(lightRender.cleanHtml).toMatchSnapshot('light mode icon');
      expect(darkRender.cleanHtml).toMatchSnapshot('dark mode icon');
    });

    it('renders correctly within light and dark themes', () => {
      // Test within light theme context
      const lightThemeRender = createComponentSnapshot(<DarkModeToggle />, { themeMode: 'light' });

      // Test within dark theme context
      const darkThemeRender = createComponentSnapshot(<DarkModeToggle />, { themeMode: 'dark' });

      expect(lightThemeRender.cleanHtml).toMatchSnapshot('within light theme');
      expect(darkThemeRender.cleanHtml).toMatchSnapshot('within dark theme');
    });
  });

  describe('Custom Classes', () => {
    it('renders with custom class names', () => {
      const { cleanHtml } = createComponentSnapshot(
        <DarkModeToggle className="custom-class-1 custom-class-2" />
      );

      expect(cleanHtml).toMatchSnapshot();
    });
  });

  describe('Accessibility', () => {
    it('has correct aria-label for light and dark modes', () => {
      // Create snapshots for both modes
      const lightAriaRender = createComponentSnapshot(<DarkModeToggle forcedMode="light" />);

      const darkAriaRender = createComponentSnapshot(<DarkModeToggle forcedMode="dark" />);

      // Verify the light mode aria-label contains the right text
      expect(lightAriaRender.cleanHtml).toContain('aria-label="Switch to dark mode"');

      // Verify the dark mode aria-label contains the right text
      expect(darkAriaRender.cleanHtml).toContain('aria-label="Switch to light mode"');

      // Save snapshots for reference
      expect(lightAriaRender.cleanHtml).toMatchSnapshot('light mode aria-label');
      expect(darkAriaRender.cleanHtml).toMatchSnapshot('dark mode aria-label');
    });
  });
});
