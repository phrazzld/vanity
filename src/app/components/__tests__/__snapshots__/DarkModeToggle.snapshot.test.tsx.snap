// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DarkModeToggle Component Snapshots Accessibility has correct aria-label for light and dark modes: dark mode aria-label 1`] = `"<div data-testid="theme-provider"><div data-theme="light"><button class="dark-mode-toggle flex items-center justify-center bg-transparent rounded-full transition-colors hover:bg-gray-100 dark:hover:bg-gray-800 p-2 " aria-label="Switch to light mode"><svg xmlns="http://www.w3.org/2000/svg" class="h-5 w-5" fill="none" viewBox="0 0 24 24" stroke="currentColor"><path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M12 3v1m0 16v1m9-9h-1M4 12H3m15.364 6.364l-.707-.707M6.343 6.343l-.707-.707m12.728 0l-.707.707M6.343 17.657l-.707.707M16 12a4 4 0 11-8 0 4 4 0 018 0z"></path></svg></button></div></div>"`;

exports[`DarkModeToggle Component Snapshots Accessibility has correct aria-label for light and dark modes: light mode aria-label 1`] = `"<div data-testid="theme-provider"><div data-theme="light"><button class="dark-mode-toggle flex items-center justify-center bg-transparent rounded-full transition-colors hover:bg-gray-100 dark:hover:bg-gray-800 p-2 " aria-label="Switch to dark mode"><svg xmlns="http://www.w3.org/2000/svg" class="h-5 w-5" fill="none" viewBox="0 0 24 24" stroke="currentColor"><path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M20.354 15.354A9 9 0 018.646 3.646 9.003 9.003 0 0012 21a9.003 9.003 0 008.354-5.646z"></path></svg></button></div></div>"`;

exports[`DarkModeToggle Component Snapshots Custom Classes renders with custom class names 1`] = `"<div data-testid="theme-provider"><div data-theme="light"><button class="dark-mode-toggle flex items-center justify-center bg-transparent rounded-full transition-colors hover:bg-gray-100 dark:hover:bg-gray-800 p-2 custom-class-1 custom-class-2" aria-label="Switch to dark mode"><svg xmlns="http://www.w3.org/2000/svg" class="h-5 w-5" fill="none" viewBox="0 0 24 24" stroke="currentColor"><path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M20.354 15.354A9 9 0 018.646 3.646 9.003 9.003 0 0012 21a9.003 9.003 0 008.354-5.646z"></path></svg></button></div></div>"`;

exports[`DarkModeToggle Component Snapshots Size Variants renders in different sizes: size: large 1`] = `"<div data-testid="theme-provider"><div data-theme="light"><button class="dark-mode-toggle flex items-center justify-center bg-transparent rounded-full transition-colors hover:bg-gray-100 dark:hover:bg-gray-800 p-2.5 " aria-label="Switch to dark mode"><svg xmlns="http://www.w3.org/2000/svg" class="h-6 w-6" fill="none" viewBox="0 0 24 24" stroke="currentColor"><path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M20.354 15.354A9 9 0 018.646 3.646 9.003 9.003 0 0012 21a9.003 9.003 0 008.354-5.646z"></path></svg></button></div></div>"`;

exports[`DarkModeToggle Component Snapshots Size Variants renders in different sizes: size: medium 1`] = `"<div data-testid="theme-provider"><div data-theme="light"><button class="dark-mode-toggle flex items-center justify-center bg-transparent rounded-full transition-colors hover:bg-gray-100 dark:hover:bg-gray-800 p-2 " aria-label="Switch to dark mode"><svg xmlns="http://www.w3.org/2000/svg" class="h-5 w-5" fill="none" viewBox="0 0 24 24" stroke="currentColor"><path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M20.354 15.354A9 9 0 018.646 3.646 9.003 9.003 0 0012 21a9.003 9.003 0 008.354-5.646z"></path></svg></button></div></div>"`;

exports[`DarkModeToggle Component Snapshots Size Variants renders in different sizes: size: small 1`] = `"<div data-testid="theme-provider"><div data-theme="light"><button class="dark-mode-toggle flex items-center justify-center bg-transparent rounded-full transition-colors hover:bg-gray-100 dark:hover:bg-gray-800 p-1.5 " aria-label="Switch to dark mode"><svg xmlns="http://www.w3.org/2000/svg" class="h-4 w-4" fill="none" viewBox="0 0 24 24" stroke="currentColor"><path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M20.354 15.354A9 9 0 018.646 3.646 9.003 9.003 0 0012 21a9.003 9.003 0 008.354-5.646z"></path></svg></button></div></div>"`;

exports[`DarkModeToggle Component Snapshots Theme Variants renders correctly within light and dark themes: within dark theme 1`] = `"<div data-testid="theme-provider"><div data-theme="dark"><button class="dark-mode-toggle flex items-center justify-center bg-transparent rounded-full transition-colors hover:bg-gray-100 dark:hover:bg-gray-800 p-2 " aria-label="Switch to dark mode"><svg xmlns="http://www.w3.org/2000/svg" class="h-5 w-5" fill="none" viewBox="0 0 24 24" stroke="currentColor"><path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M20.354 15.354A9 9 0 018.646 3.646 9.003 9.003 0 0012 21a9.003 9.003 0 008.354-5.646z"></path></svg></button></div></div>"`;

exports[`DarkModeToggle Component Snapshots Theme Variants renders correctly within light and dark themes: within light theme 1`] = `"<div data-testid="theme-provider"><div data-theme="light"><button class="dark-mode-toggle flex items-center justify-center bg-transparent rounded-full transition-colors hover:bg-gray-100 dark:hover:bg-gray-800 p-2 " aria-label="Switch to dark mode"><svg xmlns="http://www.w3.org/2000/svg" class="h-5 w-5" fill="none" viewBox="0 0 24 24" stroke="currentColor"><path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M20.354 15.354A9 9 0 018.646 3.646 9.003 9.003 0 0012 21a9.003 9.003 0 008.354-5.646z"></path></svg></button></div></div>"`;

exports[`DarkModeToggle Component Snapshots Theme Variants renders light and dark mode icons correctly: dark mode icon 1`] = `"<div data-testid="theme-provider"><div data-theme="light"><button class="dark-mode-toggle flex items-center justify-center bg-transparent rounded-full transition-colors hover:bg-gray-100 dark:hover:bg-gray-800 p-2 " aria-label="Switch to light mode"><svg xmlns="http://www.w3.org/2000/svg" class="h-5 w-5" fill="none" viewBox="0 0 24 24" stroke="currentColor"><path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M12 3v1m0 16v1m9-9h-1M4 12H3m15.364 6.364l-.707-.707M6.343 6.343l-.707-.707m12.728 0l-.707.707M6.343 17.657l-.707.707M16 12a4 4 0 11-8 0 4 4 0 018 0z"></path></svg></button></div></div>"`;

exports[`DarkModeToggle Component Snapshots Theme Variants renders light and dark mode icons correctly: light mode icon 1`] = `"<div data-testid="theme-provider"><div data-theme="light"><button class="dark-mode-toggle flex items-center justify-center bg-transparent rounded-full transition-colors hover:bg-gray-100 dark:hover:bg-gray-800 p-2 " aria-label="Switch to dark mode"><svg xmlns="http://www.w3.org/2000/svg" class="h-5 w-5" fill="none" viewBox="0 0 24 24" stroke="currentColor"><path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M20.354 15.354A9 9 0 018.646 3.646 9.003 9.003 0 0012 21a9.003 9.003 0 008.354-5.646z"></path></svg></button></div></div>"`;
